File: Midterm_example/src/main/java/formTemplateMethod/Before.java
package formTemplateMethod;

class Site{
    int unit;
    double rate;
    double TAX_RATE;

}

/**
 * ResidentialSite class and LifelineSite class
 * have the same method name but do different steps.
 */
class ResidentialSite extends Site{
    double base;
    double tax;
    double getBillableAmount(int unit,double rate){
        base = unit * rate;
        tax = base * TAX_RATE;

        return base*tax;
    }

}
class LifelineSite extends Site{
    double base;
    double tax;
    double getBillableAmount(int unit, double rate){
        base =  (unit * rate * 0.5);
        tax = base * TAX_RATE;

        return base*tax;
    }
}

File: CSharpGenelization 3/CSharpGenelization/FormTemplateMethod/Before.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharpGenelization.FormTemplateMethod
{
    internal class Before
    {
        class Site
        {
            int unit;
            double rate;
            double TAX_RATE;

            protected double getTaxRate()
            {
                return TAX_RATE;
            }
        }

        /**
         * ResidentialSite class and LifelineSite class
         * have the same method name but do different steps.
         */
        class ResidentialSite : Site
        {
            double basic;
            double tax;
            double getBillableAmount(int unit, double rate)
            {
                basic = unit * rate;
                tax = basic * base.getTaxRate();

                return basic * tax;
            }

        }
        class LifelineSite : Site
        {
            double basic;
            double tax;
            double getBillableAmount(int unit, double rate)
            {
                basic = (unit * rate * 0.5);
                tax = basic * base.getTaxRate();

                return basic * tax;
            }
        }
    }
}
